### PR TITLE
Wait for resources to resolve

### PR DESCRIPTION
**Pull request by tylerphelan**
- Builders,ClusterBuilders,ClusterStores,ClusterStacks, and import
- Helps with running commands in CI env

https://github.com/vmware-tanzu/kpack-cli/issues/50

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,7 +19,11 @@ require (
...
 	k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible
+	knative.dev/pkg v0.0.0-20200702222342-ea4d6e985ba0
 	sigs.k8s.io/yaml v1.2.0
 )
 
-replace k8s.io/client-go => k8s.io/client-go v0.17.5
+replace  (
+k8s.io/client-go => k8s.io/client-go v0.17.5
+github.com/pivotal/kpack => /Users/pivotal/workspace/kpack
+)
```
Comment (aemengo) [inline go.mod:29]: Let's remove this from the final PR

**Comment by aemengo** [comment]
I'm curious what others think.

I'm a little concerned about how much logic that this feature has introduced and how spread out it is. Perhaps, I'm being overly timorous, but I feel like this introduced more of a tech burden than is justified. Moreover I don't think of this as one of our marquee features. On the basis of cost/benefit analysis, I'm not welcome to this change given that k8s doesn't give us a cleaner API to query for resource states.

File: pkg/commands/waiter.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		kr, err := eventToDuck(&event)
+		if err != nil {
+			return false, err
+		}
+
+		if kr.Status.ObservedGeneration < generation {
+			return false, nil // still waiting on update
+		}
+
+		if kr.Status.GetCondition(apis.ConditionReady).IsUnknown() {
```
Comment (matthewmcnew) [inline pkg/commands/waiter.go:119]: I think this makes sense. Is there a time we expect this behavior. Generation observed but, unknown?

File: pkg/commands/waiter.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+type Waiter struct {
+	dynamicClient dynamic.Interface
+}
+
+func NewWaiter(dc dynamic.Interface) *Waiter {
+	return &Waiter{dynamicClient: dc}
+}
+
+func (i *Waiter) Wait(ob runtime.Object) error {
```
Comment (matthewmcnew) [inline pkg/commands/waiter.go:40]: I really like how this is generally applicable.

File: pkg/commands/waiter.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, hasResolved(m.GetObjectMeta().GetGeneration()))
+}
+
+// Used to ensure the builder has resolved to the expected stack and store (importing)
+func (i *Waiter) BuilderWait(ob runtime.Object, expectedStoreGeneration, expectedStackGeneration int64) error {
```
Comment (matthewmcnew) [inline pkg/commands/waiter.go:49]: It is a bit of a bummer that we need to have specific BuilderWait call. Could the Wait call somehow take an additional wait func or something that could just be used for builders?

File: pkg/commands/waiter.go (hunk 3 above)
Comment (matthewmcnew) [reply pkg/commands/waiter.go:40]: So good.

File: pkg/commands/waiter.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
+package commands
+
+import (
+	"context"
```
Comment (matthewmcnew) [inline pkg/commands/waiter.go:4]: Perhaps cleanup these imports?

File: pkg/commands/builder/create.go [hunk 6]
```diff
@@ -131,6 +132,9 @@ func create(name string, flags CommandFlags, ch *commands.CommandHelper, cs k8s.
 		if err != nil {
 			return err
 		}
+		if err := w.Wait(bldr); err != nil {
```
Comment (matthewmcnew) [inline pkg/commands/builder/create.go:135]: This is really clean.

**Review by matthewmcnew — APPROVED** [review]
This is really good.

File: pkg/commands/waiter.go (hunk 2 above)
Comment (tylerphelan) [reply pkg/commands/waiter.go:119]: good question! this was maybe a false assumption I made.

**Comment by tylerphelan** [comment]
Add timeouts

File: pkg/commands/waiter.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+	if err != nil {
+		return err
+	}
+	if !done {
+		m, ok := ob.(kmeta.OwnerRefable)
+		if !ok {
+			return errors.New("unexpected type")
+		}
+		name := m.GetObjectMeta().GetName()
+		gvr, _ := meta.UnsafeGuessKindToResource(m.GetGroupVersionKind())
```
Comment (sukhil-suresh) [inline pkg/commands/waiter.go:69]: assume this is not a problem for our resources.. (interesting function name 😁 )
```
// UnsafeGuessKindToResource converts Kind to a resource name.
// Broken. This method only "sort of" works when used outside of this package.  It assumes that Kinds and Resources match
// and they aren't guaranteed to do so.
func UnsafeGuessKindToResource(kind schema.GroupVersionKind) ( /*plural*/ schema.GroupVersionResource /*singular*/, schema.GroupVersionResource) {
...
```

File: pkg/commands/clusterbuilder/create.go [hunk 8]
```diff
@@ -152,6 +152,9 @@ func create(name string, flags CommandFlags, ch *commands.CommandHelper, cs k8s.
 		if err != nil {
 			return err
 		}
+		if err := waiter.Wait(cb); err != nil {
+			return err
+		}
```
Comment (sukhil-suresh) [inline pkg/commands/clusterbuilder/create.go:157]: just an observation... which applies to all the other files calling on `waiter.Wait()`
IF and when `waiter.Wait()` errors, the user is not informed of the resource object that was created.
And not sure, if we care enough to inform of resource creation & of waiting on resource to be ready separately. 

To be fair, I think this applies to `ch.PrintObj(..)` calls too

**Review by aemengo — APPROVED** [review]

File: pkg/commands/waiter.go (hunk 4 above)
Comment (tylerphelan) [reply pkg/commands/waiter.go]: updated

File: pkg/commands/waiter.go (hunk 7 above)
Comment (tylerphelan) [reply pkg/commands/waiter.go]: good question, at least we know we'll be using our resources which do have matching kinds. I've looked around and this seems to be the way to get the GroupVersionResource

File: pkg/commands/clusterbuilder/create.go (hunk 8 above)
Comment (tylerphelan) [reply pkg/commands/clusterbuilder/create.go:157]: the error will say "<resource> not ready" which I think is informative enough

**Review by sukhil-suresh — APPROVED** [review]

**Pull request merged**